File: 0126-word-ladder-ii/0126-word-ladder-ii.py
from collections import defaultdict, deque
from typing import List

class Solution:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        wordList.append(beginWord)
        wordSet = set(wordList)
        if endWord not in wordSet:
            return []

        # Adjacency list to store the graph
        adj = {}
        neigh = collections.defaultdict(list)
        for word in wordSet:
            for j in range(len(word)):
                pattern = word[:j]+"*"+word[j+1:]
                neigh[pattern].append(word)
        for word in wordSet:
            adj[word] = []
            for j in range(len(word)):
                pattern = word[:j] + "*" + word[j+1:]
                for neighbour in neigh[pattern]:
                    if neighbour != word:
                        adj[word].append(neighbour)
            
        visited = set()
        queue = deque([beginWord])
        backtrack_dict = collections.defaultdict(list)
        visited.add(beginWord)
        found = False

        while queue and not found:
            level_size = len(queue)
            current_level_visited = set()
            for _ in range(level_size):
                currentWord = queue.popleft()
                
                for word in adj[currentWord]:
                    if word == endWord:
                        found = True
                    if word not in visited:
                        if word not in current_level_visited:
                            queue.append(word)
                            current_level_visited.add(word)
                        if word != currentWord:
                            backtrack_dict[word].append(currentWord)
                            
            visited.update(current_level_visited)

        res = []
        if found:
            path = [endWord]
            self.backtrack(endWord, beginWord, backtrack_dict, path, res)
        return res

    def backtrack(self, currentWord: str, beginWord: str, adj: dict, path: List[str], res: List[List[str]]):
        if currentWord == beginWord:
            result = path.copy()
            result.reverse()
            res.append(result)
            return
        for nextWord in adj[currentWord]:
            path.append(nextWord)
            self.backtrack(nextWord, beginWord, adj, path, res)
            path.pop()

File: 0126-word-ladder-ii/0126-word-ladder-ii.cpp
class Solution {
public:
    vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> wordSet(wordList.begin(), wordList.end());
        if (wordSet.find(endWord) == wordSet.end()) return {};

        unordered_map<string, vector<string>> adj;
        unordered_set<string> visited;
        queue<string> q;
        q.push(beginWord);
        visited.insert(beginWord);
        bool found = false;

        while (!q.empty() && !found) {
            int levelSize = q.size();
            unordered_set<string> currentLevelVisited;
            for (int i = 0; i < levelSize; ++i) {
                string currentWord = q.front();
                q.pop();
                string tempWord = currentWord;

                for (int j = 0; j < currentWord.size(); ++j) {
                    char originalChar = currentWord[j];
                    for (char c = 'a'; c <= 'z'; ++c) {
                        if (c == originalChar) continue; // Skip if character is same as original
                        currentWord[j] = c;
                        if (wordSet.find(currentWord) != wordSet.end()) {
                            if (currentWord == endWord) found = true;
                            if (visited.find(currentWord) == visited.end()) {
                                if (currentLevelVisited.find(currentWord) == currentLevelVisited.end()) {
                                    q.push(currentWord);
                                    currentLevelVisited.insert(currentWord);
                                }
                                adj[currentWord].push_back(tempWord);
                            }
                        }
                    }
                    currentWord[j] = originalChar;
                }
            }
            visited.insert(currentLevelVisited.begin(), currentLevelVisited.end());
        }

        vector<vector<string>> res;
        if (found) {
            vector<string> path = {endWord};
            backtrack(endWord, beginWord, adj, path, res);
        }
        return res;
    }
    void backtrack(const string& currentWord, const string& beginWord, unordered_map<string, vector<string>>& adj, vector<string>& path, vector<vector<string>>& res) {
        if (currentWord == beginWord) {
            vector<string> result = path;
            reverse(result.begin(), result.end());
            res.push_back(result);
            return;
        }
        for (const string& nextWord : adj[currentWord]) {
            path.push_back(nextWord);
            backtrack(nextWord, beginWord, adj, path, res);
            path.pop_back();
        }
    }
            
};